Automattic/wp-calypso
# CPT: Enable trashing custom post types

Related: D2061-code, Automattic/jetpack#4180

The behavior of the [REST API `POST /sites/%s/posts/%s/delete` endpoint](https://developer.wordpress.com/docs/api/1.1/post/sites/%24site/posts/%24post_ID/delete/) is such that a post is permanently deleted if it is either (a) not a post or page or (b) already trashed. The first of these conditions led to the incorrect assumption that custom post types could not be trashed. While it is technically true that the REST API currently lacks this option, the related links above seek to add support.

The changes included here reflect the new behavior, which is to allow all post types to be saved with `trash` status. The existing `deletePost` action creator has been updated to reflect its new intended usage for permanently deleting posts (noting that if post or page type, that the post should already be trashed).

Ultimately, this simplifies behaviors since there's no longer the need for special consideration of a post's current status and type when deciding the effects of a trash.

A number of related behaviors were not previously implemented/tested, and are also included in this pull request (receiving updates for query items, delete action creator tests).

__Testing instructions:__

The `trashPost` action will currently return a `400` response code, due to lack of support for `trash` status. It's expected that this will be available shortly for WordPress.com sites and in an upcoming Jetpack release.

Test to ensure that trashing a post from the [custom post types list screen](http://calypso.localhost:3000/post), whether a post, page, or custom post type, results in that post being moved to trash, and counts updated accordingly.

__Caveats:__

- Due the changing post list item action bar to use `trashPost`, it's no longer possible to permanently delete a post with these changes. This behavior will be restored in a subsequent pull request.

/cc @timmyc 

Test live: https://calypso.live/?branch=fix/cpt-trash
